homeassistant/components/enphase_envoy/sensor.py

```diff
@@ -141,28 +207,62 @@ def device_state_attributes(self):
+        return self.coordinator.last_update_success
+
+    async def async_added_to_hass(self):
+        """When entity is added to Home Assistant."""
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    @Throttle(MIN_SCAN_INTERVAL)
     async def async_update(self):
-        """Get the energy production data from the Enphase Envoy."""
+        """Update the energy production data."""
+
+        if self.coordinator.data is None:
+            _LOGGER.debug("No data found!")
+            return
+
         if self._type != "inverters":
-            _state = await getattr(self._envoy_reader, self._type)()
-            if isinstance(_state, int):
-                self._state = _state
+            if isinstance(self.coordinator.data.get(self._type), int):
+                self._state = self.coordinator.data.get(self._type)
+                _LOGGER.debug("Updating: %s - %s", self._type, self._state)
             else:
-                _LOGGER.error(_state)
-                self._state = None
+                _LOGGER.debug(
+                    "Sensor %s isInstance(int) was %s.  Returning None for state.",
+                    self._type,
+                    isinstance(self.coordinator.data.get(self._type), int),
+                )
 
         elif self._type == "inverters":
-            try:
-                inverters = await (self._envoy_reader.inverters_production())
-            except requests.exceptions.HTTPError:
-                _LOGGER.warning(
-                    "Authentication for Inverter data failed during update: %s",
-                    self._envoy_reader.host,
+            serial_number = self._name.split(" ")[2]
+            if isinstance(self.coordinator.data.get("inverters_production"), dict):
+                self._state = self.coordinator.data.get("inverters_production").get(
+                    serial_number
+                )[0]
+                self._last_reported = self.coordinator.data.get(
+                    "inverters_production"
+                ).get(serial_number)[1]
+                _LOGGER.debug(
+                    "Updating: %s (%s) - %s.",
+                    self._type,
+                    serial_number,
+                    self._state,
                 )
-
-            if isinstance(inverters, dict):
-                serial_number = self._name.split(" ")[2]
-                self._state = inverters[serial_number][0]
-                self._last_reported = inverters[serial_number][1]
             else:
-                self._state = None
+                _LOGGER.debug(
+                    "Data inverter (%s) isInstance(dict) was %s.  Using previous state: %s",
+                    serial_number,
+                    isinstance(self.coordinator.data.get("inverters_production"), dict),
+                    self._state,
+                )
+        else:
+            _LOGGER.error("Update loop - API returned invalid data!")
```

Comment:
I can't see how you would ever reach this statement.

Reply:
Removed `else` condition